### PR TITLE
Feature/move buildings

### PR DESCRIPTION
**Pull request by Raycoms**
# Changes proposed in this pull request:
- Allow the player to move a building
- Add GUI to hut
- Add button to initital GUI
- Use GUI to move building.

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java [hunk 1]
```diff
@@ -808,7 +808,7 @@ private void changeSchematic()
      * @param complete if pasted, should it be complete.
      * @param structureName of the scan to be built.
      */
-    private static void requestScannedSchematic(@NotNull final StructureName structureName, final boolean paste, final boolean complete)
+    public static void requestScannedSchematic(@NotNull final StructureName structureName, final boolean paste, final boolean complete)
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java:811]: MAJOR Refactor this method to reduce its Cognitive Complexity from 30 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

File: src/main/java/com/minecolonies/coremod/client/gui/WindowMoveBuilding.java [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+     */
+    private void moveBackClicked()
+    {
+        Settings.instance.moveTo(new BlockPos(0, 0, 0).offset(this.mc.player.getHorizontalFacing().getOpposite()));
+    }
+
+    /**
+     * Rotate the structure clockwise.
+     */
+    private void rotateRightClicked()
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowMoveBuilding.java:225]: MINOR Make "rotateRightClicked" a "static" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2325)

File: src/main/java/com/minecolonies/coremod/client/gui/WindowMoveBuilding.java [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+    {
+        updateRotation((Settings.instance.getRotation() + ROTATE_RIGHT) % POSSIBLE_ROTATIONS);
+    }
+
+    /**
+     * Rotate the structure counter clockwise.
+     */
+    private void rotateLeftClicked()
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowMoveBuilding.java:233]: MINOR Make "rotateLeftClicked" a "static" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2325)

File: src/main/java/com/minecolonies/coremod/network/messages/BuildingMoveMessage.java [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+     * setup the building once it has been placed.
+     *  @param world         World the hut is being placed into.
+     * @param player        Who placed the hut.
+     * @param sn            The name of the structure.
+     * @param rotation      The number of times the structure should be rotated.
+     * @param buildPos      The location the hut is being placed.
+     * @param mirror        Whether or not the strcture is mirrored.
+     * @param oldBuilding The old building id.
+     */
+    private static void setupBuilding(
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/network/messages/BuildingMoveMessage.java]: MAJOR Refactor this method to reduce its Cognitive Complexity from 28 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 67 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 58 major
* ![MINOR][MINOR] 8 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [WindowBuildTool.java#L843](https://github.com/Minecolonies/minecolonies/blob/61e16bb514aa0a2eee12dcf770edfa3ba345013c/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L843): Refactor this code to not nest more than 5 if/for/while/switch/try statements. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/61e16bb514aa0a2eee12dcf770edfa3ba345013c/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/61e16bb514aa0a2eee12dcf770edfa3ba345013c/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [WindowBuildBuilding.java#L179](https://github.com/Minecolonies/minecolonies/blob/61e16bb514aa0a2eee12dcf770edfa3ba345013c/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildBuilding.java#L179): The Cyclomatic Complexity of this method "updateResources" is 22 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [WindowBuildBuilding.java#L179](https://github.com/Minecolonies/minecolonies/blob/61e16bb514aa0a2eee12dcf770edfa3ba345013c/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildBuilding.java#L179): Refactor this method to reduce its Cognitive Complexity from 35 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [WindowBuildBuilding.java#L191](https://github.com/Minecolonies/minecolonies/blob/61e16bb514aa0a2eee12dcf770edfa3ba345013c/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildBuilding.java#L191): Reduce the total number of break and continue statements in this loop to use at most one. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS135)
1. ![MAJOR][MAJOR] [WindowBuildBuilding.java#L308](https://github.com/Minecolonies/minecolonies/blob/61e16bb514aa0a2eee12dcf770edfa3ba345013c/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildBuilding.java#L308): Remove this unused method parameter "list". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1172)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L46](https://github.com/Minecolonies/minecolonies/blob/61e16bb514aa0a2eee12dcf770edfa3ba345013c/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L46): class "WindowBuildTool" has 41 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L930](https://github.com/Minecolonies/minecolonies/blob/61e16bb514aa0a2eee12dcf770edfa3ba345013c/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L930): Refactor this method to reduce its Cognitive Complexity from 24 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/61e16bb514aa0a2eee12dcf770edfa3ba345013c/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 67 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 58 major
* ![MINOR][MINOR] 8 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [WindowBuildTool.java#L843](https://github.com/Minecolonies/minecolonies/blob/4957855d77a2e02aa1f4ea058aeed32fd340e65f/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L843): Refactor this code to not nest more than 5 if/for/while/switch/try statements. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/4957855d77a2e02aa1f4ea058aeed32fd340e65f/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/4957855d77a2e02aa1f4ea058aeed32fd340e65f/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [WindowBuildBuilding.java#L180](https://github.com/Minecolonies/minecolonies/blob/4957855d77a2e02aa1f4ea058aeed32fd340e65f/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildBuilding.java#L180): The Cyclomatic Complexity of this method "updateResources" is 22 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [WindowBuildBuilding.java#L180](https://github.com/Minecolonies/minecolonies/blob/4957855d77a2e02aa1f4ea058aeed32fd340e65f/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildBuilding.java#L180): Refactor this method to reduce its Cognitive Complexity from 35 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [WindowBuildBuilding.java#L192](https://github.com/Minecolonies/minecolonies/blob/4957855d77a2e02aa1f4ea058aeed32fd340e65f/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildBuilding.java#L192): Reduce the total number of break and continue statements in this loop to use at most one. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS135)
1. ![MAJOR][MAJOR] [WindowBuildBuilding.java#L309](https://github.com/Minecolonies/minecolonies/blob/4957855d77a2e02aa1f4ea058aeed32fd340e65f/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildBuilding.java#L309): Remove this unused method parameter "list". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1172)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L46](https://github.com/Minecolonies/minecolonies/blob/4957855d77a2e02aa1f4ea058aeed32fd340e65f/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L46): class "WindowBuildTool" has 41 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L930](https://github.com/Minecolonies/minecolonies/blob/4957855d77a2e02aa1f4ea058aeed32fd340e65f/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L930): Refactor this method to reduce its Cognitive Complexity from 24 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/4957855d77a2e02aa1f4ea058aeed32fd340e65f/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Review by marchermans — APPROVED** [review]

**Review by marchermans — APPROVED** [review]

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 67 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 58 major
* ![MINOR][MINOR] 8 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [WindowBuildTool.java#L843](https://github.com/Minecolonies/minecolonies/blob/bb409f7d91af265cb40c5c01411b8a4d0b94405b/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L843): Refactor this code to not nest more than 5 if/for/while/switch/try statements. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/bb409f7d91af265cb40c5c01411b8a4d0b94405b/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/bb409f7d91af265cb40c5c01411b8a4d0b94405b/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [WindowBuildBuilding.java#L180](https://github.com/Minecolonies/minecolonies/blob/bb409f7d91af265cb40c5c01411b8a4d0b94405b/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildBuilding.java#L180): The Cyclomatic Complexity of this method "updateResources" is 22 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [WindowBuildBuilding.java#L180](https://github.com/Minecolonies/minecolonies/blob/bb409f7d91af265cb40c5c01411b8a4d0b94405b/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildBuilding.java#L180): Refactor this method to reduce its Cognitive Complexity from 35 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [WindowBuildBuilding.java#L192](https://github.com/Minecolonies/minecolonies/blob/bb409f7d91af265cb40c5c01411b8a4d0b94405b/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildBuilding.java#L192): Reduce the total number of break and continue statements in this loop to use at most one. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS135)
1. ![MAJOR][MAJOR] [WindowBuildBuilding.java#L309](https://github.com/Minecolonies/minecolonies/blob/bb409f7d91af265cb40c5c01411b8a4d0b94405b/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildBuilding.java#L309): Remove this unused method parameter "list". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1172)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L46](https://github.com/Minecolonies/minecolonies/blob/bb409f7d91af265cb40c5c01411b8a4d0b94405b/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L46): class "WindowBuildTool" has 41 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L930](https://github.com/Minecolonies/minecolonies/blob/bb409f7d91af265cb40c5c01411b8a4d0b94405b/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L930): Refactor this method to reduce its Cognitive Complexity from 24 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/bb409f7d91af265cb40c5c01411b8a4d0b94405b/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**